### PR TITLE
Provide completions for a few `builtins`

### PR DESCRIPTION
**Pull request by Ma27**
Related to #10.

I still have the dream of proper editor support for Nix expressions and this is a first, small step towards it. I have a few ideas how to proceed from here, but I can't promise that I'll find the time for further efforts on this right now.

To summarize, this PR does the following two things:

* Right now `rnix-lsp` only provides completion for symbols that can be found in the current scope. Because of that, e.g. `builtins` can't be completed properly, so I added a small mechanism which adds completions for those as well.
  Please note that the current list of builtins is incomplete, a better solution I have in mind would be to use the Nix API behind `nix __dump-builtins` from `nixUnstable`, but I need to reserve some more time on how to build a small FFI against the C++ API of Nix.
* The `lsp_types` library provides an additional field named `detail` that can be used e.g. for variable's types. I added a small layer so it's at least possible now to tell whether a lambda, a variable from a `let` block or an attribute is provided.

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -269,14 +273,15 @@ impl App {
...
-        for var in scope.keys() {
+        for (var, (datatype, _)) in scope {
             if var.starts_with(&name.as_str()) {
                 completions.push(CompletionItem {
                     label: var.clone(),
                     text_edit: Some(TextEdit {
                         range: utils::range(content, name.node().text_range()),
                         new_text: var.clone(),
                     }),
+                    detail: Some(datatype),
```
Comment (jD91mZM2) [inline src/main.rs:284]: The `datatype` variable should probably be an enum implementing Display instead, to save memory usage I guess.

**Review by jD91mZM2 — APPROVED** [review]
All in all, looks cool

File: src/main.rs (hunk 1 above)
Comment (Ma27) [reply src/main.rs]: Updated.

File: src/main.rs [hunk 2]
```diff
@@ -281,7 +281,7 @@ impl App {
                         range: utils::range(content, name.node().text_range()),
                         new_text: var.clone(),
                     }),
-                    detail: Some(datatype),
+                    detail: Some(format!("{}", datatype)),
```
Comment (jD91mZM2) [inline src/main.rs:284]: `datatype.to_string()`

File: src/utils.rs [hunk 3]
```diff
@@ -1,6 +1,28 @@
...
+
+#[derive(Clone)]
+pub enum Datatype {
+    Lambda, Variable, Attribute
+}
+
+impl Display for Datatype {
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result {
+        write!(f, "{}", match self {
+            Datatype::Lambda => "Lambda",
```
Comment (jD91mZM2) [inline src/utils.rs:14]: `Self::Lambda` if I'm being super nitpicky, sorry

File: src/utils.rs [hunk 4]
```diff
@@ -1,6 +1,28 @@
 use lsp_types::*;
 use rnix::{types::*, SyntaxNode, TextRange, TextUnit, TokenAtOffset};
 use std::{collections::HashMap, convert::TryFrom, path::PathBuf, rc::Rc};
+use std::fmt::{Display, Formatter, Result, Debug};
+
+#[derive(Clone)]
```
Comment (jD91mZM2) [inline src/utils.rs:6]: Should be `Copy`

**Review by jD91mZM2 — CHANGES_REQUESTED** [review]
Sorry for nitpicking

**Comment by jD91mZM2** [comment]
Thanks, man! You've done an awesome job at this, sorry again for being nitpicky about it. Hats off to you 🎩

**Pull request merged**